js/src/builtin/streams/ReadableByteStreamControllerOperations.cpp

```diff
@@ -0,0 +1,1237 @@
+#include "builtin/streams/QueueWithSizes.h"  // js::EnqueueValueWithSize, js::ResetQueue
+#include "builtin/streams/ReadableStreamController.h"  // js::ReadableStream{,Default}Controller, js::ReadableByteStreamController, js::ReadableStreamControllerStart{,Failed}Handler
+#include "builtin/streams/ReadableStreamInternals.h"  // js::ReadableStream{CloseInternal,ErrorInternal,FulfillReadOrReadIntoRequest,GetNumReadRequests}
+#include "builtin/streams/ReadableStreamOperations.h"  // js::ReadableStreamTee_Pull, js::SetUpReadableByteStreamController
+#include "builtin/streams/TeeState.h"  // js::TeeState
+#include "js/CallAndConstruct.h"       // JS::IsCallable
+#include "js/CallArgs.h"               // JS::CallArgs{,FromVp}
+#include "js/Promise.h"                // JS::AddPromiseReactions
+#include "js/RootingAPI.h"             // JS::Handle, JS::Rooted
+#include "js/Stream.h"                 // JS::ReadableStreamUnderlyingSource
+#include "js/Value.h"  // JS::{,Int32,Object}Value, JS::UndefinedHandleValue
+#include "vm/Compartment.h"  // JS::Compartment
+#include "vm/Interpreter.h"  // js::Call, js::GetAndClearExceptionAndStack
+#include "vm/JSContext.h"    // JSContext
+#include "vm/JSObject.h"     // JSObject
+#include "vm/List.h"         // js::ListObject
+#include "vm/PromiseObject.h"  // js::PromiseObject, js::PromiseResolvedWithUndefined
+#include "vm/Runtime.h"        // JSAtomState
+#include "vm/SavedFrame.h"  // js::SavedFrame
+
+#include "builtin/HandlerFunction-inl.h"                  // js::NewHandler
+#include "builtin/streams/MiscellaneousOperations-inl.h"  // js::PromiseCall
+#include "builtin/streams/ReadableStreamReader-inl.h"  // js::UnwrapReaderFromStreamNoThrow
+#include "vm/Compartment-inl.h"  // JS::Compartment::wrap, js::UnwrapCalleeSlot
+#include "vm/JSContext-inl.h"    // JSContext::check
+#include "vm/JSObject-inl.h"     // js::IsCallable, js::NewBuiltinClassInstance
+#include "vm/List-inl.h"         // js::ListObject, js::StoreNewListInFixedSlot
+#include "vm/Realm-inl.h"        // js::AutoRealm
+
+using js::ByteStreamChunk;
+using js::CanTransferArrayBuffer;
+using js::PullIntoDescriptor;
+using js::ReadableByteStreamController;
+using js::ReadableByteStreamControllerInvalidateBYOBRequest;
+using js::ReadableStream;
+using js::ReadableStreamController;
+using js::ReadableStreamControllerCallPullIfNeeded;
+using js::ReadableStreamControllerError;
+using js::ReadableStreamGetNumReadRequests;
+using js::ReaderType;
+using js::SavedFrame;
+using js::UnwrapCalleeSlot;
+using js::UnwrapReaderFromStreamNoThrow;
+
+using JS::CallArgs;
+using JS::CallArgsFromVp;
+using JS::Handle;
+using JS::Rooted;
+using JS::RootedValueArray;
+using JS::UndefinedHandleValue;
+using JS::Value;
+
+extern bool js::ReadableByteStreamControllerEnqueueClonedChunkToQueue(
+    JSContext* cx, JS::Handle<ReadableByteStreamController*> controller,
+    JS::Handle<JSObject*> buffer, uint32_t byteOffset, uint32_t byteLength) {
```

Comment:
Is it expected that controller or buffer could be nullptr? If not, we could add `MOZ_ASSERT`s for them both